### ISSUE TITLE
Dispatch broken for containers of objects parametrized with TypeVars

### ISSUE THREAD

**Issue by timholy** [JuliaLang/julia#12596]
Maybe I'm just not seeing it, but to me this seems to be a bug:

``` jl
abstract Paint{T, N}

abstract AbstractColor{T, N} <: Paint{T, N}
abstract Color{T}            <: AbstractColor{T, 3}
abstract Transparent{C<:AbstractColor,T,N} <: Paint{T,N}

immutable RGB{T<:FloatingPoint} <: Color{T}
    r::T # Red [0,1]
    g::T # Green [0,1]
    b::T # Blue [0,1]
end

colortype{C<:AbstractColor    }(::Type{C})                  = C
colortype{C<:AbstractColor    }(::Type{Transparent{C}})     = C
colortype{C<:AbstractColor,T  }(::Type{Transparent{C,T}})   = C
colortype{C<:AbstractColor,T,N}(::Type{Transparent{C,T,N}}) = C

T = TypeVar(:T)
colortype(RGB{Float32})                     # yields RGB{Float32}
colortype(RGB)                              # yields RGB{T<:FloatingPoint}
colortype(RGB{T})                           # yields RGB{T}
P = Transparent{RGB{Float32},Float32,4}
colortype(P)                                # yields RGB{Float32}
P = Transparent{RGB{T},T,4}

julia> colortype(P)
ERROR: LoadError: MethodError: `colortype` has no method matching colortype(::Type{Transparent{RGB{T},T,4}})
 in include at ./boot.jl:254
 in include_from_node1 at ./loading.jl:264
while loading /tmp/color_trouble2.jl, in expression starting on line 25
```

Yet

``` jl
C = TypeVar(:C,AbstractColor)
N = TypeVar(:N)

julia> typeintersect(P, Transparent{C,T,N})
Transparent{RGB{T},T,4}

julia> P <: Transparent{C,T,N}
true
```

**Comment by timholy** [comment]
It's actually the 2nd parameter that's causing the problem, and it depends on whether it's a bound or unbound typevar. Reduced case:
```jl
abstract AbstractColor{T, N}
abstract Color{T}            <: AbstractColor{T, 3}
abstract Transparent{C<:AbstractColor,T,N}

immutable RGB{T<:FloatingPoint} <: Color{T}
    r::T # Red [0,1]
    g::T # Green [0,1]
    b::T # Blue [0,1]
end

C = TypeVar(:C,AbstractColor,true)
Tu = TypeVar(:T,false)
Tb = TypeVar(:T,true)
N = TypeVar(:N,true)
B = Transparent{C,Tb,N}

julia> P = Transparent{RGB{Tu},Tu,4}
Transparent{RGB{T},T,4}

julia> typeintersect(Type{P}, Type{B})
Union{}

julia> P = Transparent{RGB{Tu},Tb,4}
Transparent{RGB{T},T,4}

julia> typeintersect(Type{P}, Type{B})
Type{Transparent{RGB{T},T,4}}
```
The key lines seem to be [here](https://github.com/JuliaLang/julia/blob/72c344c472bfd55fcda36aa30585c8f185389cdd/src/jltypes.c#L513-L515).

This may give me enough guidance to work around this.

**Comment by mbauman** [comment]
Sounds like https://github.com/JuliaLang/julia/issues/2552#issuecomment-120096696 and/or https://github.com/JuliaLang/julia/issues/11597.  I've actually stopped restricting type parameters in my own code and just enforcing it by what constructors I provide because of wonkiness like this and because I often want "triangular" constraints which aren't expressible in any case.

**Comment by timholy** [comment]
Yeah, this indeed popped up in a rework of Color.jl, for which I'm trying to design some nice traits to compensate for the lack of triangular dispatch. It specifically arose through introspection on `TypeConstructor`s which arise from `typealias`es. I got around it by ditching all the typealiases and generating types by metaprogramming.

**Comment by timholy** [comment]
Rearing its head again. Here's a simple real-world example:
```jl
julia> abstract A{T,N}

julia> immutable B{T} <: A{T,3}
           val::T
       end

julia> len{T,N}(::Type{A{T,N}}) = N
len (generic function with 1 method)

julia> super(B)
A{T,3}

julia> len(super(B))
ERROR: MethodError: `len` has no method matching len(::Type{A{T,3}})
```
This seems to be a consequence of fixing #7062. I have to say, this seems like a worse problem to have than #7062. What do you think, @vtjnash?

**Comment by timholy** [comment]
Just to show you how devious I am (in case you had any remaining doubts), here's a workaround:
```jl
julia> len2{N}(::Type{A{TypeVar(:T),N}}) = N
len2 (generic function with 1 method)

julia> len2(super(B))
3
```

**Comment by vtjnash** [comment]
> https://github.com/JuliaLang/julia/issues/12596#issuecomment-131334565

that answer looks correct to me. the TypeVar `T` is not being matched in the first example, so the method cannot be used. the official way to avoid this is to use a typealias (although constructing the TypeVar manually is pretty similar).

**Comment by timholy** [comment]
I wonder whether this issue should turn into "further disambiguate the printing of method tables and types so the proper dispatch rules are obvious"?

**Comment by JeffBezanson** [comment]
closed by #18457
